cmd/thanos/query.go

```diff
@@ -271,6 +302,46 @@ func runQuery(
 			stores.Close()
 		})
 	}
+	// Run File Service Discovery and update the store set when the files are modified.
+	if fileSD != nil {
+		var fileSDUpdates chan []*targetgroup.Group
+		ctxRun, cancelRun := context.WithCancel(context.Background())
+
+		fileSDUpdates = make(chan []*targetgroup.Group)
+
+		g.Add(func() error {
+			fileSD.Run(ctxRun, fileSDUpdates)
+			return nil
+		}, func(error) {
+			cancelRun()
+		})
+
+		ctxUpdate, cancelUpdate := context.WithCancel(context.Background())
+		g.Add(func() error {
+			for {
+				select {
+				case update, ok := <-fileSDUpdates:
+					// Handle the case that a discoverer exits and closes the channel
+					// before the context is done.
+					if !ok {
+						return nil
+					}
+					// Discoverers sometimes send nil updates so need to check for it to avoid panics.
+					if update == nil {
+						continue
+					}
+					// TODO(ivan): resolve dns here maybe?
+					fileSDCache.update(update)
+					stores.Update(ctxUpdate)
+				case <-ctxUpdate.Done():
+					return nil
+				}
+			}
+		}, func(error) {
+			cancelUpdate()
+			close(fileSDUpdates)
```

Comment:
hm.. Can we ensure we close this ONLY in one place? I feel like both `fileSD` and this guy closes this, right?

Reply:
File SD doesn't close it. It is responsibility of the one who passes the channel to file SD.

Reply:
so why this comment exists? 
```
// Handle the case that a discoverer exits and closes the channel
					// before the context is done.
```

Reply:
yeah, good question. I think this is not needed then. I will remove it. I double checked again but file sd doesn't close this channel.

Reply:
to clarify, the check when receiving from the channel is not needed.